src/appleseed/renderer/modeling/camera/fisheyelenscamera.cpp

```diff
@@ -0,0 +1,428 @@
+                m_film_dimensions[1],
+                m_focal_length,
+                m_near_z,
+                m_shutter_open_begin_time,
+                m_shutter_open_end_time,
+                m_shutter_close_begin_time,
+                m_shutter_close_end_time,
+                projection_type);
+        }
+
+        bool on_render_begin(
+            const Project&          project,
+            const BaseGroup*        parent,
+            OnRenderBeginRecorder&  recorder,
+            IAbortSwitch*           abort_switch) override
+        {
+            if (!PerspectiveCamera::on_render_begin(project, parent, recorder, abort_switch))
+                return false;
+
+            const string projection_type = m_params.get_required<string>("projection_type", "equisolid_angle");
+
+            if (projection_type == "equisolid_angle")
+                 m_projection_type = Projection::EquisolidAngle;
+             else if (projection_type == "equidistant")
+                 m_projection_type = Projection::Equidistant;
+             else if (projection_type == "stereographic")
+                 m_projection_type = Projection::Stereographic;
+             else if (projection_type == "thoby")
+                 m_projection_type = Projection::Thoby;
+             else
+             {
+                 RENDERER_LOG_ERROR(
+                     "invalid value \"%s\" for parameter \"projection_type\", "
+                     "using default value \"equisolid_angle\".",
+                     projection_type.c_str());
+                 m_projection_type = Projection::EquisolidAngle;
+             }
+
+            return true;
+        }
+
+        void spawn_ray(
+            SamplingContext&        sampling_context,
+            const Dual2d&           ndc,
+            ShadingRay&             ray) const override
+        {
+            // Initialize the ray.
+            initialize_ray(sampling_context, ray);
+
+            // Retrieve the camera transform.
+            Transformd scratch;
+            const Transformd& transform =
+                m_transform_sequence.evaluate(ray.m_time.m_absolute, scratch);
+
+            // Compute ray origin and direction.
+            ray.m_org = transform.get_local_to_parent().extract_translation();
+            ray.m_dir = normalize(transform.vector_to_parent(-ndc_to_camera(ndc.get_value())));
+
+            // Compute ray derivatives.
+            if (ndc.has_derivatives())
+            {
+                const Vector2d px(ndc.get_value() + ndc.get_dx());
+                const Vector2d py(ndc.get_value() + ndc.get_dy());
+                ray.m_rx.m_org = ray.m_org;
+                ray.m_ry.m_org = ray.m_org;
+                ray.m_rx.m_dir = normalize(transform.vector_to_parent(-ndc_to_camera(px)));
+                ray.m_ry.m_dir = normalize(transform.vector_to_parent(-ndc_to_camera(py)));
+                ray.m_has_differentials = true;
+            }
+        }
+
+        bool connect_vertex(
```

Comment:
Here is what I had written at the time when you opened this PR but that I hadn't posted (because I was still thinking about it):

> > connect_vertex() is same with Pinhole Camera's one. Compared to Pinhole renders, fisheye renders brightness seems good to me. So I thought connect_vertex() works fine, but I'm not sure.. appreciate for feedbacks.
> 
> I think that brightness differences between a correct implementation of `connect_vertex()` and an incorrect one will be hard to see with the naked eye, except maybe with extremely strong distortions.
> 
> So in a sense, we could live with this implementation. However it's still desirable to have a correct implementation if it's not probihibitive or overly complex to compute. It's also a nice exercise.
> 
> One way to approach the problem is this: a camera is a function that maps 2D coordinates in the image plane to 3D rays (3D origin and 3D direction, so technically a 6D domain).
> 
> For instance, an orthographic camera maps 2D image plane coordinates to parallel rays. A spherical camera maps 2D image plane coordinates to rays with identical origins but with directions covering the whole sphere, but it doesn't do that uniformly: it's the classical problem of mapping a rectangle onto a sphere, there are poles (discontinuities) as well as areas of higher or lower density. The importance returned by `connect_vertex()` must be proportional to this density of rays.
> 
> As an example, check how the spherical camera works: it first computes the outgoing direction vector: it's just a vector from the camera's origin to the vertex to connect. It then computes the NDC (image plane) coordinates that correspond to this direction. It then considers a small patch of 1x1 pixel centered on the 2D image plane sample and maps it back onto a sphere. Finally it computes the solid angle sustained by that small spherical patch. That's what it returns as importance.